### DISCUSSION TITLE
Even Asphere Sag Depth

### DISCUSSION THREAD

**Discussion by jakecarr3** [mjhoptics/ray-optics#152] [Q&A] [answered]
I've been working with a few aspheric surfaces and noticed the aspheric profile didn't look right in the pictures. So I checked the sag depth and it doesn't match. 

I used the sag depth calculator [here](https://wavelength-oe.com/optical-calculators/aspheric-surface/) to verify. 

I've been entering the coefficients like an even asphere A2, A4, A6, A8, etc. I tried adding extra 0s in the coefficients like if it took even and odd terms with A1, A2, A3, A4 etc. terms, but the sag depth is unchanged. If I zero out all the coefficients and just leave the cc, the sag value reported stays the same too. Only if I change the cc, the sag depth updates so I think it's somehow not getting the asphere terms. 
 
```
opm.radius_mode = True
sm.do_apertures = False
```

```
sm.add_surface([95.867658, 5.08, 2.6563, 45.72])
sm.ifcs[sm.cur_surface].profile = RadialPolynomial(r=95.867658, cc=0, coefs=[0., 8.180497E-7, -7.007436E-12,  -5.576895E-15, 4.688910E-18])
print(sm.ifcs[sm.cur_surface].profile.sag(45.72, 0.))
```
Output 
`11.604442713583566`

Should be 15.19550

![image](https://github.com/user-attachments/assets/38a4bc54-18a9-43bc-a91a-0f65bda9914f)

Another Example

```
sm.add_surface([48.358836, -5.08, 'refl', 36.575])
sm.ifcs[sm.cur_surface].profile = RadialPolynomial(r=48.358836, cc=-2.474466, coefs=[0., 3.460039E-6, 1.053395E-10, 2.273815E-13, 1.045178E-16])
print(sm.ifcs[sm.cur_surface].profile.sag(36.575, 0.))
```
Output 
`11.73271321584345`

Should be 19.35262

![image](https://github.com/user-attachments/assets/18b9beb2-1663-4983-8732-ea98727a2501)

**Comment by mjhoptics** [comment]
Hello @jakecarr3,
Thank you for your question. The `profile` class that you want to use is the EvenPolynomial, rather than the RadialPolynomial. The RadialPolynomial profile includes odd terms as well as even (e.g. A1, A3, A5, etc.). I think the doc could be clearer on this; I'll resolve the issue when I've done that.
Thanks for the clear report.
Mike Hayford


```python
from rayoptics.environment import *
```


```python
opm = OpticalModel()
sm  = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']
```


```python
opm.radius_mode = True
sm.do_apertures = False
sm.gaps[0].thi = 1e11
```

# Radial vs Even Polynomials

## RadialPolynomial profile

The [RadialPolynomial](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.elem.profiles.html#rayoptics.elem.profiles.RadialPolynomial) profile includes both even and odd terms in r. The entries for rad_poly correspond to the 2nd, 3rd, 4th, and 5th powers of r.

Note that creating standalone profile instances requires you to explicitly call the update() method on the profile instance. 


```python
rad_poly = RadialPolynomial(r=95.867658, cc=0, coefs=[0., 8.180497E-7, -7.007436E-12,  -5.576895E-15, 4.688910E-18])
rad_poly.update()
print(f"radial poly sag = {rad_poly.sag(45.72, 0.):12.5f}")
```

    radial poly sag =     11.60615


## EvenPolynomial profile

The [EvenPolynomial](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.elem.profiles.html#rayoptics.elem.profiles.EvenPolynomial) profile include only even powers of r. The entries for even_poly correspond to the  4th, 6th, 8th, and 10th powers of r.

The update() method for the profile must be called for the sag() evaluation to be correct.


```python
even_poly = EvenPolynomial(r=95.867658, cc=0, coefs=[0., 8.180497E-7, -7.007436E-12,  -5.576895E-15, 4.688910E-18])
even_poly.update()
print(f"even poly sag = {even_poly.sag(45.72, 0.):12.5f}")
```

    even poly sag =     15.19550


The RadialPolynomial may be used as long as the odd polynomial terms are explicitly entered as zero.


```python
rad_poly1 =  RadialPolynomial(r=95.867658, cc=0, coefs=[0., 0., 0., 8.180497E-7, 0., -7.007436E-12,  0., -5.576895E-15, 0., 4.688910E-18])
rad_poly1.update()
print(f"radial poly 1 sag = {rad_poly1.sag(45.72, 0.):12.5f}")
```

    radial poly 1 sag =     15.19550


## Use of update_model()

Normally, when entering profiles as part of the model definition, the call to update_model() on the optical model will update all of the profiles in the model, as seen below.


```python
sm.add_surface([95.867658, 5.08, 2.6563], sd=45.72)
rad_poly2 =  RadialPolynomial(r=95.867658, cc=0, coefs=[0., 0., 0., 8.180497E-7, 0., -7.007436E-12,  0., -5.576895E-15, 0., 4.688910E-18])
sm.ifcs[sm.cur_surface].profile = rad_poly2
```


```python
sm.add_surface([95.867658, 5.08, 2.6563], sd=45.72)
even_poly1 = EvenPolynomial(r=95.867658, cc=0, coefs=[0., 8.180497E-7, -7.007436E-12,  -5.576895E-15, 4.688910E-18])
sm.ifcs[sm.cur_surface].profile = even_poly1
```


```python
opm.update_model()
```


```python
listobj(even_poly1)
print(f"even poly 1 sag = {even_poly1.sag(45.72, 0.):12.5f}")
```

    profile: EvenPolynomial
    c=0.010431046516229696,   r=95.867658   conic cnst=0
    coefficients: [0.0, 8.180497e-07, -7.007436e-12, -5.576895e-15, 4.68891e-18]
    
    even poly 1 sag =     15.19550



```python
listobj(rad_poly2)
print(f"radial poly 2 sag = {rad_poly2.sag(45.72, 0.):12.5f}")
```

    profile: RadialPolynomial
    c=0.010431046516229696,   r=95.867658   conic cnst=0.0
    coefficients: [0.0, 0.0, 0.0, 8.180497e-07, 0.0, -7.007436e-12, 0.0, -5.576895e-15, 0.0, 4.68891e-18]
    
    radial poly 2 sag =     15.19550


# Example 2, using the EvenPolynomial


```python
example2 = EvenPolynomial(r=48.358836, cc=-2.474466, coefs=[0., 3.460039E-6, 1.053395E-10, 2.273815E-13, 1.045178E-16])
example2.update()

print(f"example2 sag = {example2.sag(36.575, 0.):12.5f}")
```

    example2 sag =     19.35262



```python

```

**Comment by jakecarr3** [comment, marked as answer]
Even Polynominal and update for the win! That worked perfectly. Thanks for your help and all the work you've done putting this together.
